### ISSUE TITLE
Transform gain function

### ISSUE THREAD

**Issue by parsiad** [microsoft/LightGBM#4073]
When splitting a leaf on a given predictor, we pick the split to maximize the gain function (e.g., see the equation under "Learn the Tree Structure" at https://xgboost.readthedocs.io/en/latest/tutorials/model.html#learn-the-tree-structure from the XGBoost docs).

I would like to modify the gain function "globally". To be more precise, denoting the gain function `gain_func(split_point)`, I would like to construct a new function `new_gain_func = T[gain_func]` and pick the split by maximizing `new_gain_func(split_point)`.

Does something like this exist already and if not, do you have any recommendations for how to implement it?

Thanks in advance.

**Comment by shiyu1994** [comment]
@parsiad Thanks for using LightGBM. We're sorry that currently LightGBM doesn't support transformation on split gains.

To implement this, you just need to modify the `GetSplitGains` function in `src/treelearner/feature_histogram.hpp`. The function is used to calculate all the gains, given the summation of gradients and hessians of two child nodes of a split. I think it should easy to do such a modification.

If you have any further problems, please feel free to post here.

**Comment by parsiad** [comment]
@shiyu1994: Thank you for your response! I will try to modify GetSplitGains as you suggest.

**Comment by parsiad** [comment]
@shiyu1994: I am returning to this issue after a small hiatus. It seems like `GetSplitGains` computes the gain of a *single* split threshold and that I really need to inject my logic into `FindBestThresholdSequentially`, as that will compute the gain of *multiple* split thresholds.

For my functionality, I need the actual value of the split threshold. It seems like in this function, a split threshold is stored as an index corresponding to a bin:
```cpp
best_threshold = static_cast<uint32_t>(t - 1 + offset);
```
Do you know how to go from the index to the bin value in this function?

**Comment by parsiad** [comment]
I ended up using [BinToValue](https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/include/LightGBM/bin.h#L114). Closing this thread.

**Comment by shiyu1994** [comment]
@parsiad Sorry for the late response. I was busy in previous days. Yes `BinToValue` would be a correct option to do that. If in your code, you have access to `Dataset` object, you may also use its `RealThreshold` method. See
https://github.com/microsoft/LightGBM/blob/b8a6f809e9dcfd8560099a8c3df493afcd2fbe94/include/LightGBM/dataset.h#L568-L572
